File: PolyPaint_Server/src/Models/Canvas.ts
import { Base, IBase } from "./Base"


export enum CAP {
    SQUARE = 0,
    ROUND = 1
}

/**
 * Canvas Interfaces
 */
export interface IPosition {
    _id: string,
    x: number,
    y: number
}

export interface IColor {
    _id: string,
    r: number,
    g: number,
    b: number,
    a: number
}

export interface IStroke extends IBase {
    _id: string,
    canvas_id: string,
    size: number,
    color: IColor,
    cap: CAP,
    coordinates: IPosition[],
    toDelete?: boolean
}

export interface ICanvas extends IBase {
    _id: string,
    strokes: IStroke[]
    uids: string[] // TEMP: For testing only
}

/**
 * Canvas Classes
 */
export class Stroke extends Base implements IStroke, IBase {
    constructor(
        public _id: string,
        public canvas_id: string,
        public size: number,
        public color: IColor,
        public cap: CAP,
        public coordinates: IPosition[],
        public toDelete: boolean = false,
    ){
        super()
    }
}

export class Canvas extends Base implements ICanvas, IBase {
    constructor(
        public _id: string,
        public strokes: IStroke[],
        public uids: string[] // TEMP: For testing only
    ){
        super()
    }
}

File: PolyPaint_ClientLourd/PolyPaint/Models/Game/Canvas.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PolyPaint.Models.Game
{

    public struct Coordinate
    {
        [JsonProperty(PropertyName = "x")]
        public double X { get; set; }
        [JsonProperty(PropertyName = "y")]
        public double Y { get; set; }
    }
    


    public class Path
    {
        [JsonProperty(PropertyName = "_id")]
        public string ID { get; set; }
        [JsonProperty(PropertyName = "canvas_id")]
        public string CanvasID { get; set; }
        [JsonProperty(PropertyName = "size")]
        public int Size { get; set; }
        [JsonProperty(PropertyName = "color")]
        public string Color { get; set; }
        [JsonProperty(PropertyName = "coordinates")]
        public  List<Coordinate> Coordinates { get; set; }
        [JsonProperty(PropertyName = "toDelete")]
        public bool ToDelete { get; set; } = false;
        [JsonProperty(PropertyName = "cap")]
        public Cap DrawingShape { get; set; }

        public Path()
        {
            Coordinates = new List<Coordinate>();
        }

       public Path(string id, string canvasID, int size, string color, bool toDelete, string drawingShape)
        {
            ID = id;
            CanvasID = canvasID;
            Size = size;
            Color = color;
            Coordinates = new List<Coordinate>();
            ToDelete = toDelete;
            DrawingShape = drawingShape == "ronde" ? Cap.Round : Cap.Square;
        }

        public Path DeepCopy()
        {
            var drawingShape = DrawingShapeToString();
            var copy = new Path(ID, CanvasID, Size, Color, ToDelete, drawingShape);
            copy.Coordinates = new List<Coordinate>(Coordinates);

            return copy;
        }

        private string DrawingShapeToString()
        {
            return DrawingShape == Cap.Round ? "ronde" : "carree";
        }
    }

   

    public class Canvas
    {
        [JsonProperty(PropertyName = "_id", Required=Required.AllowNull)]
        public string ID { get; set; }
        [JsonProperty(PropertyName = "strokes", Required = Required.AllowNull)]
        public List<Path> Strokes { get; set; }
        [JsonProperty(PropertyName = "uids", Required = Required.AllowNull)]
        public List<string> UserIds  { get; set; }

        public Canvas()
        {
            ID = string.Empty;
            Strokes = new List<Path>();
            UserIds = new List<string>();
        }

        
    }
    public enum Cap
    {
        Square,
        Round
    }

}
